sklearn/preprocessing/data.py

```diff
@@ -1811,138 +1813,248 @@ class OneHotEncoder(BaseEstimator, TransformerMixin):
-        indices = np.cumsum(n_values)
-        self.feature_indices_ = indices
+            categorical = np.zeros(n_features, dtype=bool)
+            categorical[np.asarray(self.categorical_features)] = True
+
+        num_cat = np.sum(categorical)
+        start = 0
+        cat_index = 0
+
+        for i in range(n_features):
+            if categorical[i]:
+                le = self._label_encoders[cat_index]
+                end = start + len(le.classes_)
+                self.feature_index_range_[i] = start, end
+                start = end
+                cat_index += 1
+
+        indices = np.arange(start, start + n_features - num_cat)
+        self.feature_index_range_[~categorical, 0] = indices
+        indices += 1
+        self.feature_index_range_[~categorical, 1] = indices
+
+        if len(indices) > 0:
+            output_cols = indices[-1]
+        else:
+            output_cols = start
 
-        column_indices = (X + indices[:-1]).ravel()
-        row_indices = np.repeat(np.arange(n_samples, dtype=np.int32),
-                                n_features)
-        data = np.ones(n_samples * n_features)
-        out = sparse.coo_matrix((data, (row_indices, column_indices)),
-                                shape=(n_samples, indices[-1]),
-                                dtype=self.dtype).tocsr()
+        self.one_hot_feature_index_ = np.empty(output_cols, dtype=np.int)
 
-        if (isinstance(self.n_values, six.string_types) and
-                self.n_values == 'auto'):
-            mask = np.array(out.sum(axis=0)).ravel() != 0
-            active_features = np.where(mask)[0]
-            out = out[:, active_features]
-            self.active_features_ = active_features
+        for i in range(n_features):
+            s, e = self.feature_index_range_[i]
+            self.one_hot_feature_index_[s:e] = i
 
-        return out if self.sparse else out.toarray()
+        return self
 
-    def fit_transform(self, X, y=None):
-        """Fit OneHotEncoder to X, then transform X.
+    def _fit(self, X):
+        "Assumes `X` contains only catergorical features."
 
-        Equivalent to self.fit(X).transform(X), but more convenient and more
-        efficient. See fit for the parameters, transform for the return value.
+        if not np.issubdtype(X.dtype.type, np.integer):
+            X = check_array(X, dtype=np.object)
+        n_samples, n_features = X.shape
+
+        self._n_features = n_features
+        self._label_encoders = [LabelEncoder() for i in range(n_features)]
+        # Maximum value for each featue
+        self._max_values = [None for i in range(n_features)]
+
+        if self.n_values is not None:
+            warnings.warn('The parameter `n_values` is deprecated, use the'
+                          'parameter `values` instead and specify the '
+                          'expected values for each feature')
+
+            if isinstance(self.n_values, numbers.Integral):
```

Comment:
Why do we need to repeat this logic, not just `self.values = self.n_values` and change the warning to "`n_values` has been renamed to `values`" so that the later messages are interpretable...